### PR TITLE
Simplify display of Union and Tuple types by creation of FoldingBuilder

### PR DESCRIPTION
**Pull request by danimontoya**
Draft that covers part of #499 , Union and Tuple types.
### Union Types
- Simplify display of Union2<A, B> to A | B when displayed in the editor and back to edit mode when focused
### Tuple Types
- Simplify display of Tuple2<A, B> to (A, B) when displayed in the editor and back to edit mode when focused

### CODE REVIEW

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/foldingbuilder/FoldingSyntax.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      object : KeyedLazyInstance<FoldingBuilder> {
+        override fun getKey(): String = lang.displayName
+
+        override fun getInstance(): FoldingBuilder = builder
+      }
+    )
+
+  fun IdeMetaPlugin.addFoldingBuilder(
+    placeHolderText: (node: ASTNode) -> String?,
+    foldRegions: (node: ASTNode, document: Document) -> Array<FoldingDescriptor>,
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/foldingbuilder/FoldingSyntax.kt:30]:
```suggestion
    foldRegions: (node: ASTNode, document: Document) -> List<FoldingDescriptor>,
```
lists are simpler :)

**Comment by i-walker** [comment]
We can merge the algebra already if you tested it and it works.

**Comment by danimontoya** [comment]
WIP - I haven't tested it yet.. its just the code you shared yesterday 👍

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/utils/IdeUtils.kt [hunk 2]
```diff
@@ -24,13 +24,7 @@ import org.jetbrains.kotlin.idea.caches.resolve.analyzeWithAllCompilerChecks
...
 import org.jetbrains.kotlin.idea.fir.firResolveState
 import org.jetbrains.kotlin.idea.fir.getOrBuildFir
-import org.jetbrains.kotlin.psi.KtCallElement
-import org.jetbrains.kotlin.psi.KtCallableDeclaration
-import org.jetbrains.kotlin.psi.KtElement
-import org.jetbrains.kotlin.psi.KtFile
-import org.jetbrains.kotlin.psi.KtNamedFunction
-import org.jetbrains.kotlin.psi.KtProperty
-import org.jetbrains.kotlin.psi.KtPsiFactory
+import org.jetbrains.kotlin.psi.*
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/utils/IdeUtils.kt:27]: use explicit imports

File: idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/utils/IdeUtils.kt [hunk 3]
```diff
@@ -53,6 +47,9 @@ fun <A : PsiElement, B> PsiElement.traverseFilter(on: Class<A>, f: (A) -> B): Li
 val KtElement.callElements: List<KtCallElement>
   get() = traverseFilter(KtCallElement::class.java) { it }
 
+val KtElement.typeReferences: List<KtTypeReference>
+  get() = traverseFilter(KtTypeReference::class.java) { it }
```
Comment (i-walker) [inline idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/utils/IdeUtils.kt:51]: this is `sequence` update to the latest commit in type-proofs branch

**Comment by i-walker** [comment]
Thanks @danimontoya 🎉 !

**Pull request merged**